File: frontend/src/types/trash.ts
/**
 * Trash domain types (Governance & Administration - Phase 2).
 *
 * Trash is a **read + restore** surface for soft-deleted records. There is no
 * permanent-delete shape here - that is deferred to RBAC. Catalog values match
 * the backend `StrEnum`s; the UI translates them only for display.
 */

import {
  ASSET_STATUSES,
  ASSET_TYPES,
  CRITICALITIES,
  ENVIRONMENTS,
  type AssetStatus,
  type AssetType,
  type Criticality,
  type Environment,
} from "@/types/asset";
import {
  INCIDENT_PRIORITIES,
  INCIDENT_SEVERITIES,
  INCIDENT_STATUSES,
  type IncidentEvent,
  type IncidentPriority,
  type IncidentSeverity,
  type IncidentStatus,
} from "@/types/incident";

/** The "who / when" every trashed record carries. */
interface DeletedMeta {
  deleted_at: string;
  deleted_by: string | null;
  deleted_by_email: string | null;
}

export interface TrashAssetListItem extends DeletedMeta {
  id: string;
  name: string;
  asset_type: AssetType;
  environment: Environment;
  criticality: Criticality;
  status: AssetStatus;
}

export interface TrashAssetDetail extends DeletedMeta {
  id: string;
  name: string;
  asset_type: AssetType;
  environment: Environment;
  criticality: Criticality;
  status: AssetStatus;
  hostname: string | null;
  ip_address: string | null;
  owner: string | null;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TrashIncidentListItem extends DeletedMeta {
  id: string;
  title: string;
  severity: IncidentSeverity;
  status: IncidentStatus;
  priority: IncidentPriority;
  owner: string | null;
  affected_asset_count: number;
}

export interface TrashIncidentAssetRef {
  id: string;
  name: string;
  asset_type: string;
  environment: string;
  criticality: string;
  status: string;
  is_active: boolean;
  deleted_at: string | null;
}

export interface TrashIncidentDetail extends DeletedMeta {
  id: string;
  title: string;
  description: string | null;
  severity: IncidentSeverity;
  status: IncidentStatus;
  priority: IncidentPriority;
  owner: string | null;
  started_at: string;
  detected_at: string | null;
  resolved_at: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  affected_assets: TrashIncidentAssetRef[];
  timeline: IncidentEvent[];
}

export interface TrashAssetPage {
  items: TrashAssetListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface TrashIncidentPage {
  items: TrashIncidentListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface TrashSummary {
  assets: number;
  incidents: number;
}

// --- runtime guards (the network response is untrusted) --------------

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function oneOf<T extends readonly string[]>(list: T, v: unknown): v is T[number] {
  return typeof v === "string" && (list as readonly string[]).includes(v);
}

function isAssetCatalog(v: Record<string, unknown>): boolean {
  return (
    oneOf(ASSET_TYPES, v.asset_type) &&
    oneOf(ENVIRONMENTS, v.environment) &&
    oneOf(CRITICALITIES, v.criticality) &&
    oneOf(ASSET_STATUSES, v.status)
  );
}

function isIncidentCatalog(v: Record<string, unknown>): boolean {
  return (
    oneOf(INCIDENT_SEVERITIES, v.severity) &&
    oneOf(INCIDENT_STATUSES, v.status) &&
    oneOf(INCIDENT_PRIORITIES, v.priority)
  );
}

function hasDeletedMeta(v: Record<string, unknown>): boolean {
  return (
    typeof v.deleted_at === "string" &&
    (v.deleted_by === null || typeof v.deleted_by === "string") &&
    (v.deleted_by_email === null || typeof v.deleted_by_email === "string")
  );
}

export function isTrashAssetListItem(value: unknown): value is TrashAssetListItem {
  if (!isRecord(value)) return false;
  return (
    typeof value.id === "string" &&
    typeof value.name === "string" &&
    isAssetCatalog(value) &&
    hasDeletedMeta(value)
  );
}

export function isTrashAssetPage(value: unknown): value is TrashAssetPage {
  if (!isRecord(value)) return false;
  return Array.isArray(value.items) && value.items.every(isTrashAssetListItem) && isPageMeta(value);
}

export function isTrashIncidentListItem(value: unknown): value is TrashIncidentListItem {
  if (!isRecord(value)) return false;
  return (
    typeof value.id === "string" &&
    typeof value.title === "string" &&
    isIncidentCatalog(value) &&
    typeof value.affected_asset_count === "number" &&
    hasDeletedMeta(value)
  );
}

export function isTrashIncidentPage(value: unknown): value is TrashIncidentPage {
  if (!isRecord(value)) return false;
  return (
    Array.isArray(value.items) && value.items.every(isTrashIncidentListItem) && isPageMeta(value)
  );
}

export function isTrashAssetDetail(value: unknown): value is TrashAssetDetail {
  if (!isRecord(value)) return false;
  return (
    typeof value.id === "string" &&
    typeof value.name === "string" &&
    isAssetCatalog(value) &&
    typeof value.is_active === "boolean" &&
    typeof value.created_at === "string" &&
    hasDeletedMeta(value)
  );
}

export function isTrashIncidentDetail(value: unknown): value is TrashIncidentDetail {
  if (!isRecord(value)) return false;
  return (
    typeof value.id === "string" &&
    typeof value.title === "string" &&
    isIncidentCatalog(value) &&
    Array.isArray(value.affected_assets) &&
    Array.isArray(value.timeline) &&
    typeof value.created_by === "string" &&
    hasDeletedMeta(value)
  );
}

export function isTrashSummary(value: unknown): value is TrashSummary {
  if (!isRecord(value)) return false;
  return typeof value.assets === "number" && typeof value.incidents === "number";
}

function isPageMeta(v: Record<string, unknown>): boolean {
  return (
    typeof v.page === "number" &&
    typeof v.page_size === "number" &&
    typeof v.total === "number" &&
    typeof v.total_pages === "number"
  );
}

File: backend/app/schemas/trash.py
"""Request/response schemas for the Trash API (Governance & Administration - Phase 2).

Trash is a **dedicated read path** for soft-deleted records. Normal Asset /
Incident schemas are reused where possible; the extra ``deleted_at`` /
``deleted_by`` / ``deleted_by_email`` fields only ever appear here.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.asset import AssetStatus, AssetType, Criticality, Environment
from app.models.incident import IncidentPriority, IncidentSeverity, IncidentStatus
from app.schemas.incident import IncidentEventRead

# Trash pages are lighter than the operational lists; keep them scannable.
ASSETS_DEFAULT_PAGE_SIZE = 20
INCIDENTS_DEFAULT_PAGE_SIZE = 15
MAX_PAGE_SIZE = 100


class _DeletedMeta(BaseModel):
    """The 'who / when' every trashed row carries."""

    deleted_at: datetime
    deleted_by: uuid.UUID | None
    deleted_by_email: str | None


# --- Assets --------------------------------------------------------------


class TrashAssetListItem(_DeletedMeta):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    asset_type: AssetType
    environment: Environment
    criticality: Criticality
    status: AssetStatus


class TrashAssetDetail(_DeletedMeta):
    """Full read-only projection of a trashed asset - everything the operator
    needs to decide whether to restore it."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    asset_type: AssetType
    environment: Environment
    criticality: Criticality
    status: AssetStatus
    hostname: str | None
    ip_address: str | None
    owner: str | None
    description: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class TrashAssetPage(BaseModel):
    items: list[TrashAssetListItem]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=MAX_PAGE_SIZE)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


# --- Incidents ---------------------------------------------------------


class TrashIncidentListItem(_DeletedMeta):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    severity: IncidentSeverity
    status: IncidentStatus
    priority: IncidentPriority
    owner: str | None
    affected_asset_count: int = Field(ge=0)


class TrashIncidentAssetRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    asset_type: AssetType
    environment: Environment
    criticality: Criticality
    status: AssetStatus
    is_active: bool
    deleted_at: datetime | None = None


class TrashIncidentDetail(_DeletedMeta):
    """Full read-only projection of a trashed incident, with its **preserved**
    timeline and affected-asset relationships."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    description: str | None
    severity: IncidentSeverity
    status: IncidentStatus
    priority: IncidentPriority
    owner: str | None
    started_at: datetime
    detected_at: datetime | None
    resolved_at: datetime | None
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    affected_assets: list[TrashIncidentAssetRef]
    timeline: list[IncidentEventRead]


class TrashIncidentPage(BaseModel):
    items: list[TrashIncidentListItem]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=MAX_PAGE_SIZE)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


# --- Summary ----------------------------------------------------------


class TrashSummary(BaseModel):
    """Compact counters for the Trash header - a recovery workspace, not a
    dashboard."""

    assets: int = Field(ge=0)
    incidents: int = Field(ge=0)
